File: examples/inputs/deps/nexus-temporal-types/go/temporal_model_converters.go
package model

import (
	"fmt"
	"reflect"
	"time"

	common "go.temporal.io/api/common/v1"
	deploymentpb "go.temporal.io/api/deployment/v1"
	enums "go.temporal.io/api/enums/v1"
	taskqueue "go.temporal.io/api/taskqueue/v1"
	workflowpb "go.temporal.io/api/workflow/v1"
	"go.temporal.io/sdk/client"
	"go.temporal.io/sdk/converter"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
	"google.golang.org/protobuf/types/known/durationpb"
)

// --- Duration (google.protobuf.Duration) ---

func durationToProto(_ workflow.Context, d *time.Duration) (*durationpb.Duration, error) {
	if d == nil || *d == 0 {
		return nil, nil
	}
	return durationpb.New(*d), nil
}

func durationFromProto(_ workflow.Context, d *durationpb.Duration) (*time.Duration, error) {
	if d == nil {
		return nil, nil
	}
	value := d.AsDuration()
	return &value, nil
}

// --- TaskQueue (temporal.api.taskqueue.v1.TaskQueue) ---

func taskQueueToProto(_ workflow.Context, name *string) (*taskqueue.TaskQueue, error) {
	if name == nil {
		return nil, nil
	}
	return &taskqueue.TaskQueue{Name: *name}, nil
}

func taskQueueFromProto(_ workflow.Context, tq *taskqueue.TaskQueue) (*string, error) {
	if tq == nil {
		return nil, nil
	}
	value := tq.GetName()
	return &value, nil
}

// --- RetryPolicy (temporal.api.common.v1.RetryPolicy) ---

func retryPolicyToProto(_ workflow.Context, p *temporal.RetryPolicy) (*common.RetryPolicy, error) {
	if p == nil || retryPolicyIsZero(*p) {
		return nil, nil
	}
	proto := &common.RetryPolicy{
		BackoffCoefficient:     p.BackoffCoefficient,
		MaximumAttempts:        p.MaximumAttempts,
		NonRetryableErrorTypes: p.NonRetryableErrorTypes,
	}
	if p.InitialInterval != 0 {
		proto.InitialInterval = durationpb.New(p.InitialInterval)
	}
	if p.MaximumInterval != 0 {
		proto.MaximumInterval = durationpb.New(p.MaximumInterval)
	}
	return proto, nil
}

func retryPolicyIsZero(p temporal.RetryPolicy) bool {
	return p.InitialInterval == 0 &&
		p.BackoffCoefficient == 0 &&
		p.MaximumInterval == 0 &&
		p.MaximumAttempts == 0 &&
		len(p.NonRetryableErrorTypes) == 0
}

func retryPolicyFromProto(_ workflow.Context, p *common.RetryPolicy) (*temporal.RetryPolicy, error) {
	if p == nil {
		return nil, nil
	}
	policy := temporal.RetryPolicy{
		BackoffCoefficient:     p.GetBackoffCoefficient(),
		MaximumAttempts:        p.GetMaximumAttempts(),
		NonRetryableErrorTypes: p.GetNonRetryableErrorTypes(),
	}
	if interval := p.GetInitialInterval(); interval != nil {
		policy.InitialInterval = interval.AsDuration()
	}
	if interval := p.GetMaximumInterval(); interval != nil {
		policy.MaximumInterval = interval.AsDuration()
	}
	return &policy, nil
}

// --- Priority (temporal.api.common.v1.Priority) ---

func priorityToProto(_ workflow.Context, p *temporal.Priority) (*common.Priority, error) {
	if p == nil || *p == (temporal.Priority{}) {
		return nil, nil
	}
	return &common.Priority{
		PriorityKey:    int32(p.PriorityKey),
		FairnessKey:    p.FairnessKey,
		FairnessWeight: p.FairnessWeight,
	}, nil
}

func priorityFromProto(_ workflow.Context, p *common.Priority) (*temporal.Priority, error) {
	if p == nil {
		return nil, nil
	}
	return &temporal.Priority{
		PriorityKey:    int(p.GetPriorityKey()),
		FairnessKey:    p.GetFairnessKey(),
		FairnessWeight: p.GetFairnessWeight(),
	}, nil
}

// --- WorkflowType (temporal.api.common.v1.WorkflowType) ---

func workflowTypeToProto(_ workflow.Context, name *string) (*common.WorkflowType, error) {
	if name == nil {
		return nil, nil
	}
	return &common.WorkflowType{Name: *name}, nil
}

func workflowTypeFromProto(_ workflow.Context, t *common.WorkflowType) (*string, error) {
	if t == nil {
		return nil, nil
	}
	value := t.GetName()
	return &value, nil
}

// --- Payload / Payloads (temporal.api.common.v1.Payload[s]) ---
func payloadToProto(ctx workflow.Context, value any) (*common.Payload, error) {
	return getWorkflowDataConverter(ctx).ToPayload(value)
}

func payloadFromProto(ctx workflow.Context, payload *common.Payload) (any, error) {
	if payload == nil {
		return nil, nil
	}
	var value any
	if err := getWorkflowDataConverter(ctx).FromPayload(payload, &value); err != nil {
		return nil, err
	}
	return value, nil
}

func payloadsToProto(ctx workflow.Context, values []any) (*common.Payloads, error) {
	if len(values) == 0 {
		return nil, nil
	}
	payloads, err := getWorkflowDataConverter(ctx).ToPayloads(values...)
	if err != nil {
		return nil, err
	}
	return payloads, nil
}

func payloadsFromProto(ctx workflow.Context, payloads *common.Payloads) ([]any, error) {
	if payloads == nil {
		return nil, nil
	}
	values := make([]any, 0, len(payloads.GetPayloads()))
	for _, payload := range payloads.GetPayloads() {
		value, err := payloadFromProto(ctx, payload)
		if err != nil {
			return nil, err
		}
		values = append(values, value)
	}
	return values, nil
}

func getWorkflowDataConverter(ctx workflow.Context) converter.DataConverter {
	dataConverter := converter.GetDefaultDataConverter()
	if options := ctx.Value("wfEnvOptions"); options != nil {
		optionsValue := reflect.ValueOf(options)
		if optionsValue.Kind() == reflect.Pointer && !optionsValue.IsNil() {
			optionsValue = optionsValue.Elem()
		}
		if optionsValue.Kind() == reflect.Struct {
			field := optionsValue.FieldByName("DataConverter")
			if field.IsValid() && field.CanInterface() && !field.IsNil() {
				if value, ok := field.Interface().(converter.DataConverter); ok {
					dataConverter = value
				}
			}
		}
	}
	if contextAware, ok := dataConverter.(workflow.ContextAware); ok {
		return contextAware.WithWorkflowContext(ctx)
	}
	return dataConverter
}

// --- Memo (temporal.api.common.v1.Memo) ---

func memoToProto(ctx workflow.Context, memo map[string]any) (*common.Memo, error) {
	if memo == nil {
		return nil, nil
	}
	fields := make(map[string]*common.Payload, len(memo))
	for key, value := range memo {
		payload, err := payloadToProto(ctx, value)
		if err != nil {
			return nil, fmt.Errorf("encode workflow memo error: %v", err)
		}
		fields[key] = payload
	}
	return &common.Memo{Fields: fields}, nil
}

func memoFromProto(ctx workflow.Context, memo *common.Memo) (map[string]any, error) {
	if memo == nil {
		return nil, nil
	}
	result := make(map[string]any, len(memo.GetFields()))
	for key, payload := range memo.GetFields() {
		value, err := payloadFromProto(ctx, payload)
		if err != nil {
			return nil, err
		}
		result[key] = value
	}
	return result, nil
}

// --- SearchAttributes (temporal.api.common.v1.SearchAttributes) ---

func searchAttributesToProto(_ workflow.Context, searchAttributes *temporal.SearchAttributes) (*common.SearchAttributes, error) {
	if searchAttributes == nil || searchAttributes.Size() == 0 {
		return nil, nil
	}

	fields := make(map[string]*common.Payload, searchAttributes.Size())
	for key, value := range searchAttributes.GetUntypedValues() {
		payload, err := converter.GetDefaultDataConverter().ToPayload(value)
		if err != nil {
			return nil, fmt.Errorf("encode search attribute [%s] error: %v", key, err)
		}
		if payload.GetData() != nil {
			if payload.Metadata == nil {
				payload.Metadata = map[string][]byte{}
			}
			payload.Metadata["type"] = []byte(key.GetValueType().String())
		}
		fields[key.GetName()] = payload
	}
	return &common.SearchAttributes{IndexedFields: fields}, nil
}

// --- VersioningOverride (temporal.api.workflow.v1.VersioningOverride) ---

func versioningOverrideToProto(_ workflow.Context, versioningOverride *client.VersioningOverride) (*workflowpb.VersioningOverride, error) {
	if versioningOverride == nil || *versioningOverride == nil {
		return nil, nil
	}

	switch v := (*versioningOverride).(type) {
	case *client.PinnedVersioningOverride:
		return &workflowpb.VersioningOverride{
			Behavior:      enums.VERSIONING_BEHAVIOR_PINNED,
			PinnedVersion: v.Version.DeploymentName + "." + v.Version.BuildID,
			Deployment: &deploymentpb.Deployment{
				SeriesName: v.Version.DeploymentName,
				BuildId:    v.Version.BuildID,
			},
			Override: &workflowpb.VersioningOverride_Pinned{
				Pinned: &workflowpb.VersioningOverride_PinnedOverride{
					Behavior: workflowpb.VersioningOverride_PINNED_OVERRIDE_BEHAVIOR_PINNED,
					Version: &deploymentpb.WorkerDeploymentVersion{
						DeploymentName: v.Version.DeploymentName,
						BuildId:        v.Version.BuildID,
					},
				},
			},
		}, nil
	case *client.AutoUpgradeVersioningOverride:
		return &workflowpb.VersioningOverride{
			Behavior: enums.VERSIONING_BEHAVIOR_AUTO_UPGRADE,
			Override: &workflowpb.VersioningOverride_AutoUpgrade{
				AutoUpgrade: true,
			},
		}, nil
	default:
		return nil, nil
	}
}

// --- Workflow context options (sourced fields) ---

// WorkflowContextOptions configures a workflow started from workflow code.
type WorkflowContextOptions struct {
	// Namespace is the namespace in which to start or signal the workflow.
	// The current workflow's namespace is used when this is empty.
	Namespace string
	// ID is the required workflow ID.
	ID string
	// TaskQueue is the required task queue for a newly started workflow.
	TaskQueue string
	// WorkflowExecutionTimeout is the end-to-end timeout, including retries and continue-as-new.
	WorkflowExecutionTimeout time.Duration
	// WorkflowRunTimeout is the timeout for a single workflow run.
	WorkflowRunTimeout time.Duration
	// WorkflowTaskTimeout is the timeout for a single workflow task.
	WorkflowTaskTimeout time.Duration
	// WorkflowIDReusePolicy controls reuse of an ID from a closed workflow.
	WorkflowIDReusePolicy enums.WorkflowIdReusePolicy
	// RetryPolicy configures retries for the workflow.
	RetryPolicy *temporal.RetryPolicy
	// CronSchedule starts the workflow on the given cron schedule.
	CronSchedule string
	// Memo is the non-indexed information attached to the workflow.
	Memo map[string]any
	// SearchAttributes contains the typed search attributes attached to the workflow.
	SearchAttributes temporal.SearchAttributes
	// Priority configures the workflow's task priority.
	Priority temporal.Priority
}

type workflowContextOptionsKey struct{}

// WithWorkflowContextOptions returns a context carrying options for starting a workflow.
func WithWorkflowContextOptions(ctx workflow.Context, options WorkflowContextOptions) workflow.Context {
	return workflow.WithValue(ctx, workflowContextOptionsKey{}, options)
}

func workflowContextOptions(ctx workflow.Context) WorkflowContextOptions {
	options, _ := ctx.Value(workflowContextOptionsKey{}).(WorkflowContextOptions)
	return options
}

func workflowStartWorkflowID(ctx workflow.Context) string {
	id := workflowContextOptions(ctx).ID
	if id == "" {
		panic("workflow ID is required in WorkflowContextOptions")
	}
	return id
}

func workflowStartTaskQueue(ctx workflow.Context) string {
	taskQueue := workflowContextOptions(ctx).TaskQueue
	if taskQueue == "" {
		panic("task queue is required in WorkflowContextOptions")
	}
	return taskQueue
}

func workflowStartExecutionTimeout(ctx workflow.Context) time.Duration {
	return workflowContextOptions(ctx).WorkflowExecutionTimeout
}

func workflowStartRunTimeout(ctx workflow.Context) time.Duration {
	return workflowContextOptions(ctx).WorkflowRunTimeout
}

func workflowStartTaskTimeout(ctx workflow.Context) time.Duration {
	return workflowContextOptions(ctx).WorkflowTaskTimeout
}

func workflowStartIDReusePolicy(ctx workflow.Context) enums.WorkflowIdReusePolicy {
	return workflowContextOptions(ctx).WorkflowIDReusePolicy
}

func workflowStartRetryPolicy(ctx workflow.Context) temporal.RetryPolicy {
	if retryPolicy := workflowContextOptions(ctx).RetryPolicy; retryPolicy != nil {
		return *retryPolicy
	}
	return temporal.RetryPolicy{}
}

func workflowStartCronSchedule(ctx workflow.Context) string {
	return workflowContextOptions(ctx).CronSchedule
}

func workflowStartMemo(ctx workflow.Context) map[string]any {
	return workflowContextOptions(ctx).Memo
}

func workflowStartSearchAttributes(ctx workflow.Context) temporal.SearchAttributes {
	return workflowContextOptions(ctx).SearchAttributes
}

func workflowStartPriority(ctx workflow.Context) temporal.Priority {
	return workflowContextOptions(ctx).Priority
}

func workflowStartNamespace(ctx workflow.Context) string {
	namespace := workflowContextOptions(ctx).Namespace
	if namespace != "" {
		return namespace
	}
	return workflowNamespace(ctx)
}

// --- Workflow namespace (sourced field) ---

func workflowNamespace(ctx workflow.Context) string {
	if options := ctx.Value("wfEnvOptions"); options != nil {
		optionsValue := reflect.ValueOf(options)
		if optionsValue.Kind() == reflect.Pointer && !optionsValue.IsNil() {
			optionsValue = optionsValue.Elem()
		}
		if optionsValue.Kind() == reflect.Struct {
			field := optionsValue.FieldByName("Namespace")
			if field.IsValid() && field.Kind() == reflect.String {
				return field.String()
			}
		}
	}
	return ""
}

File: examples/inputs/deps/nexus-temporal-types/python/temporal_model_converters.py
# pyright: reportUnusedParameter=false

import collections.abc
from datetime import timedelta
import typing

import google.protobuf.duration_pb2
import temporalio.api.common.v1.message_pb2 as common_pb2
import temporalio.api.enums.v1.workflow_pb2 as workflow_enums_pb2
import temporalio.api.taskqueue.v1.message_pb2 as taskqueue_pb2
import temporalio.api.workflow.v1
import temporalio.converter
import temporalio.common


def _current_payload_converter(
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> temporalio.converter.PayloadConverter:
    if payload_converter is not None:
        return payload_converter

    from temporalio.workflow import payload_converter as workflow_payload_converter

    return workflow_payload_converter()


def retry_policy_from_proto(
    proto: common_pb2.RetryPolicy,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> temporalio.common.RetryPolicy:
    return temporalio.common.RetryPolicy.from_proto(proto)


def retry_policy_to_proto(
    retry_policy: temporalio.common.RetryPolicy,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> common_pb2.RetryPolicy:
    proto = common_pb2.RetryPolicy()
    retry_policy.apply_to_proto(proto)
    return proto


def workflow_function_name(
    value: str | collections.abc.Callable[..., collections.abc.Awaitable[object]],
) -> str:
    from temporalio.workflow import _Definition  # pyright: ignore[reportPrivateUsage]

    name, _result_type = _Definition.get_name_and_result_type(value)
    return name


def signal_function_to_proto(
    value: str | collections.abc.Callable[..., typing.Any],
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> str:
    from temporalio.workflow import _SignalDefinition  # pyright: ignore[reportPrivateUsage]

    return _SignalDefinition.must_name_from_fn_or_str(value)  # pyright: ignore[reportUnknownMemberType]


def workflow_type_to_proto(
    workflow_type: str
    | collections.abc.Callable[..., collections.abc.Awaitable[object]],
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> common_pb2.WorkflowType:
    return common_pb2.WorkflowType(name=workflow_function_name(workflow_type))


def workflow_type_from_proto(
    proto: common_pb2.WorkflowType,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> str:
    return proto.name


def task_queue_from_proto(
    proto: taskqueue_pb2.TaskQueue,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> str:
    return proto.name


def task_queue_to_proto(
    task_queue: str,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> taskqueue_pb2.TaskQueue:
    return taskqueue_pb2.TaskQueue(name=task_queue)


def workflow_namespace() -> str:
    from temporalio.workflow import info

    return info().namespace


def payloads_to_proto(
    values: collections.abc.Sequence[typing.Any],
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> common_pb2.Payloads:
    return _current_payload_converter(payload_converter).to_payloads_wrapper(values)


def payloads_from_proto(
    proto: common_pb2.Payloads,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> list[object]:
    return list(_current_payload_converter(payload_converter).from_payloads_wrapper(proto))


def _clone_payload(payload: common_pb2.Payload) -> common_pb2.Payload:
    clone = common_pb2.Payload()
    clone.CopyFrom(payload)
    return clone


def _value_to_payload(
    value: object | common_pb2.Payload,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> common_pb2.Payload:
    if isinstance(value, common_pb2.Payload):
        return _clone_payload(value)

    payloads = _current_payload_converter(payload_converter).to_payloads_wrapper([value])
    return _clone_payload(payloads.payloads[0])


def _payload_to_value(
    payload: common_pb2.Payload,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> object:
    wrapper = common_pb2.Payloads()
    wrapper.payloads.add().CopyFrom(payload)

    return typing.cast(
        object,
        _current_payload_converter(payload_converter).from_payloads_wrapper(wrapper)[0],
    )


def payload_from_proto(
    proto: common_pb2.Payload,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> object:
    return _payload_to_value(proto, payload_converter)


def payload_to_proto(
    payload: object,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> common_pb2.Payload:
    return _value_to_payload(payload, payload_converter)


def memo_from_proto(
    proto: common_pb2.Memo,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> collections.abc.Mapping[str, object]:
    return {
        key: _payload_to_value(value, payload_converter)
        for key, value in proto.fields.items()
    }


def memo_to_proto(
    memo: collections.abc.Mapping[str, object],
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> common_pb2.Memo:
    message = common_pb2.Memo()
    for key, value in memo.items():
        message.fields[key].CopyFrom(_value_to_payload(value, payload_converter))
    return message


def duration_from_proto(
    proto: google.protobuf.duration_pb2.Duration,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> timedelta:
    return proto.ToTimedelta()


def duration_to_proto(
    duration: timedelta,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> google.protobuf.duration_pb2.Duration:
    proto = google.protobuf.duration_pb2.Duration()
    proto.FromTimedelta(duration)
    return proto


def workflow_id_reuse_policy_from_proto(
    policy: workflow_enums_pb2.WorkflowIdReusePolicy.ValueType,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> temporalio.common.WorkflowIDReusePolicy:
    return temporalio.common.WorkflowIDReusePolicy(int(policy))


def workflow_id_reuse_policy_to_proto(
    policy: temporalio.common.WorkflowIDReusePolicy,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> workflow_enums_pb2.WorkflowIdReusePolicy.ValueType:
    return typing.cast(workflow_enums_pb2.WorkflowIdReusePolicy.ValueType, int(policy))


def workflow_id_conflict_policy_from_proto(
    policy: workflow_enums_pb2.WorkflowIdConflictPolicy.ValueType,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> temporalio.common.WorkflowIDConflictPolicy:
    return temporalio.common.WorkflowIDConflictPolicy(int(policy))


def workflow_id_conflict_policy_to_proto(
    policy: temporalio.common.WorkflowIDConflictPolicy,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> workflow_enums_pb2.WorkflowIdConflictPolicy.ValueType:
    return typing.cast(
        workflow_enums_pb2.WorkflowIdConflictPolicy.ValueType, int(policy)
    )


def search_attributes_to_proto(
    search_attributes: temporalio.common.TypedSearchAttributes,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> common_pb2.SearchAttributes:
    proto = common_pb2.SearchAttributes()
    temporalio.converter.encode_search_attributes(search_attributes, proto)
    return proto


def search_attributes_from_proto(
    proto: common_pb2.SearchAttributes,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> temporalio.common.TypedSearchAttributes:
    return temporalio.converter.decode_typed_search_attributes(proto)


def priority_from_proto(
    proto: common_pb2.Priority,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> temporalio.common.Priority:
    return temporalio.common.Priority._from_proto(proto)  # pyright: ignore[reportPrivateUsage]


def priority_to_proto(
    priority: temporalio.common.Priority,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> common_pb2.Priority:
    return priority._to_proto()  # pyright: ignore[reportPrivateUsage]


def versioning_override_to_proto(
    versioning_override: temporalio.common.VersioningOverride,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> temporalio.api.workflow.v1.VersioningOverride:
    return versioning_override._to_proto()  # pyright: ignore[reportPrivateUsage]


def versioning_override_from_proto(
    proto: temporalio.api.workflow.v1.VersioningOverride,
    *,
    payload_converter: temporalio.converter.PayloadConverter | None = None,
) -> temporalio.common.VersioningOverride:
    if proto.HasField("pinned") and proto.pinned.HasField("version"):
        version = proto.pinned.version
        return temporalio.common.PinnedVersioningOverride(
            temporalio.common.WorkerDeploymentVersion(
                deployment_name=version.deployment_name,
                build_id=version.build_id,
            )
        )
    if proto.pinned_version:
        return temporalio.common.PinnedVersioningOverride(
            temporalio.common.WorkerDeploymentVersion.from_canonical_string(
                proto.pinned_version
            )
        )
    if proto.auto_upgrade:
        return temporalio.common.AutoUpgradeVersioningOverride()
    raise ValueError("unknown versioning override proto shape")
